core/executor/src/wasm_executor.rs

```diff
@@ -50,18 +50,52 @@ struct FunctionExecutor<'e, E: Externalities<Blake2Hasher> + 'e> {
 	heap: allocator::FreeingBumpHeapAllocator,
 	memory: MemoryRef,
 	table: Option<TableRef>,
-	ext: &'e mut E,
+	ext: ExtHideout<'e, E>,
 	hash_lookup: HashMap<Vec<u8>, Vec<u8>>,
 }
 
+/// Hides externalities reference so that it is hard to reference externalities directly.
+struct ExtHideout<'e, E: Externalities<Blake2Hasher> + 'e>(&'e mut E);
+
+impl<'e, E: Externalities<Blake2Hasher> + 'e> ExtHideout<'e, E> {
+	/// Execute closure that has access to externalities.
+	///
+	/// All panics that happen within closure **are NOT** captured.
+	pub fn with_ext<T, F>(&mut self, f: F) -> Result<T>
+		where
+			F: FnOnce(&mut dyn Externalities<Blake2Hasher>) -> Result<T>
+	{
+		f(&mut *self.0)
+	}
+
+	/// Execute closure that has access to storage externalities part.
+	///
+	/// All panics that happen within closure are captured and transformed into
+	/// runtime error. This requires special panic handler mode to be enabled
+	/// during the call (see `panic_handler::AbortGuard::never_abort`).
+	/// If this mode isn't enabled, then all panics within externalities are
+	/// leading to process abort.
+	pub fn with_storage_ext<T, F>(&mut self, f: F) -> Result<T>
+		where
+			F: ::std::panic::UnwindSafe + FnOnce(&mut dyn StorageExternalities<Blake2Hasher>) -> Result<T>
```

Comment:
This line exceed 100 chars which is acceptable but we can still improve it:

1. I believe that we don't need the leading `::` 
2. Since there are a few `std::` doesn't it warrant us to have `use std::panic`?